Enable fast switching among models at the invoke> command line

lstein opened the pull request:
# New commands:

This PR enables two new commands in the invoke.py script
- `!models`              -- list the available models and their cache status
- `!switch <model>` -- switch to the indicated model
    
Example:
~~~
invoke> !models
laion400m            not loaded  Latent Diffusion LAION400M model
stable-diffusion-1.4     active  Stable Diffusion inference model version 1.4
waifu-1.3                cached  Waifu anime model version 1.3
invoke> !switch waifu-1.3
>> Caching model stable-diffusion-1.4 in system RAM
>> Retrieving model waifu-1.3 from system RAM cache
~~~

The name and descriptions of the models are taken from `config/models.yaml`. A future enhancement to `model_cache.py` will be to enable new model stanzas to be added to the file programmatically. This will be useful for the WebGUI.

# More details:
    
- Use fast switching algorithm described in PR #948, implemented in new module `ldm/invoke/model_cache.py`
- Models are selected using their configuration stanza name given in models.yaml.
- To avoid filling up CPU RAM with cached models, this PR implements an LRU cache that monitors available CPU RAM.
- The caching code allows the minimum value of available RAM to be adjusted, but invoke.py does not currently have a command-line argument that allows you to set it. The minimum free RAM is arbitrarily set to 2 GB.
- Add optional description field to configs/models.yaml
    
# Unrelated fixes:
- Added ">>" to CompViz model loading messages in order to make user experience more consistent.
- When generating an image greater than defaults, will only warn about possible VRAM filling the first time.
- Fixed bug that was causing help message to be printed twice. This involved moving the import line for the web backend into the section where it is called.

Co-authored by @ArDiouscuros

lstein commented:
Hey folks, I've tested the fast model switching on both a CUDA system and an intel box with no GPU, so I think it will work on the Mac, but I'm asking @Any-Winter-4079 to check it out just in case.

ArDiouscuros commented on ldm/generate.py:
```diff
@@ -635,29 +646,45 @@ def _make_inpaint(self):
+
+        self._set_sampler()
+
+        for m in self.model.modules():
```
Imo, it should be done only once after load from file. 
If m.padding_mode was already changed and we switch to the model it will replace original one.

lstein replied:
Right. I'll move that bit of code to the file load.

lstein replied:
Code is moved to only execute once in latest commit.

Any-Winter-4079 commented:
I have a uni project for tomorrow, so I'll be mostly unavailable today. So, if you want a quick review maybe @Vargol or @skurovec can chime in

skurovec commented:
Tested on M1 Max  32GB.  

stable diffusion 1.4
```"banana sushi" -s 50 -S 1760350044 -W 512 -H 512 -C 8.0 -A ddim```

image generated


![000583 1760350044](https://user-images.githubusercontent.com/2119229/195357850-a5341054-a596-47e1-8bcb-7dc12f351a8a.png)


```!switch waifu-diffusion-1.3```
```"banana sushi" -s 50 -S 1760350044 -W 512 -H 512 -C 8.0 -A ddim```

image generated


![000584 1760350044](https://user-images.githubusercontent.com/2119229/195357903-43d280cf-1e9f-40ce-ba25-9deb0f1f35d3.png)


```!switch waifu-diffusion-1.2```
```"banana sushi" -s 50 -S 1760350044 -W 512 -H 512 -C 8.0 -A ddim```

image generated

![000585 1760350044](https://user-images.githubusercontent.com/2119229/195358034-f7cf8efe-a2d5-48bd-aa6b-1cdf7849a83d.png)


```!switch stable-diffusion-1.4```
```"banana sushi" -s 50 -S 1760350044 -W 512 -H 512 -C 8.0 -A ddim```

crashed with:

```
RuntimeError: Placeholder storage has not been allocated on MPS device!
```

Switching on another cached model throws same error.

skurovec commented:
It look like it is because in model_cache it is not returned to mps, only for cuda.

```    
def _model_from_cpu(self,model):
        if self._has_cuda():
            model.to(self.device)
            model.first_stage_model.to(self.device)
            model.cond_stage_model.to(self.device)
            model.cond_stage_model.device = self.device

        return model
```

If I add ```model.to(self.device)```  it is workign again and generatign same image.

```
    def _model_from_cpu(self,model):
        if self._has_cuda():
            model.to(self.device)
            model.first_stage_model.to(self.device)
            model.cond_stage_model.to(self.device)
            model.cond_stage_model.device = self.device
        else:
            model.to(self.device)

        return model
```

lstein commented:
H'mmm. On second thought, I'm not sure whether this will fix the problem.
On non-CUDA systems, the only device is "cpu", and so the caching commands
are essentially intended to be no-ops on MPS systems.

Unless I'm fundamentally misunderstanding something!

Lincoln

On Wed, Oct 12, 2022 at 10:14 AM Lincoln Stein ***@***.***>
wrote:

> Stupid bug! I'll fix.
>
> Lincoln
>
> On Wed, Oct 12, 2022 at 9:58 AM Jan Skurovec ***@***.***>
> wrote:
>
>> It look like it is because in model_cache it is not returned to mps, only
>> for cuda.
>>
>> def _model_from_cpu(self,model):
>>         if self._has_cuda():
>>             model.to(self.device)
>>             model.first_stage_model.to(self.device)
>>             model.cond_stage_model.to(self.device)
>>             model.cond_stage_model.device = self.device
>>
>>         return model
>>
>> If I add model.to(self.device) it is workign again and generatign same
>> image.
>>
>>     def _model_from_cpu(self,model):
>>         if self._has_cuda():
>>             model.to(self.device)
>>             model.first_stage_model.to(self.device)
>>             model.cond_stage_model.to(self.device)
>>             model.cond_stage_model.device = self.device
>>         else:
>>             model.to(self.device)
>>
>>         return model
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/invoke-ai/InvokeAI/pull/1066#issuecomment-1276225199>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAA3EVMH4QC52NOB4RETDGLWC27ZPANCNFSM6AAAAAARC6U4ZI>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
> --
> Written on my cell phone. Anything that seems odd is the fault of
> auto-correct.
>
-- 
Written on my cell phone. Anything that seems odd is the fault of
auto-correct.

lstein commented:
Or is there an "mps" device?

On Wed, Oct 12, 2022 at 10:24 AM Lincoln Stein ***@***.***>
wrote:

> H'mmm. On second thought, I'm not sure whether this will fix the problem.
> On non-CUDA systems, the only device is "cpu", and so the caching commands
> are essentially intended to be no-ops on MPS systems.
>
> Unless I'm fundamentally misunderstanding something!
>
> Lincoln
>
> On Wed, Oct 12, 2022 at 10:14 AM Lincoln Stein ***@***.***>
> wrote:
>
>> Stupid bug! I'll fix.
>>
>> Lincoln
>>
>> On Wed, Oct 12, 2022 at 9:58 AM Jan Skurovec ***@***.***>
>> wrote:
>>
>>> It look like it is because in model_cache it is not returned to mps,
>>> only for cuda.
>>>
>>> def _model_from_cpu(self,model):
>>>         if self._has_cuda():
>>>             model.to(self.device)
>>>             model.first_stage_model.to(self.device)
>>>             model.cond_stage_model.to(self.device)
>>>             model.cond_stage_model.device = self.device
>>>
>>>         return model
>>>
>>> If I add model.to(self.device) it is workign again and generatign same
>>> image.
>>>
>>>     def _model_from_cpu(self,model):
>>>         if self._has_cuda():
>>>             model.to(self.device)
>>>             model.first_stage_model.to(self.device)
>>>             model.cond_stage_model.to(self.device)
>>>             model.cond_stage_model.device = self.device
>>>         else:
>>>             model.to(self.device)
>>>
>>>         return model
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/invoke-ai/InvokeAI/pull/1066#issuecomment-1276225199>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AAA3EVMH4QC52NOB4RETDGLWC27ZPANCNFSM6AAAAAARC6U4ZI>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> ***@***.***>
>>>
>> --
>> Written on my cell phone. Anything that seems odd is the fault of
>> auto-correct.
>>
> --
> Written on my cell phone. Anything that seems odd is the fault of
> auto-correct.
>
-- 
Written on my cell phone. Anything that seems odd is the fault of
auto-correct.

skurovec commented:
```self.device.type``` returns 'mps'

lstein commented:
So the code should be:

def _model_from_cpu(self,model):
        if self.device != 'cpu':
            model.to(self.device)
            model.first_stage_model.to(self.device)
            model.cond_stage_model.to(self.device)
            model.cond_stage_model.device = self.device

        return model

I'm at a conference right now, but will commit this fix sometime today.


L


On Wed, Oct 12, 2022 at 10:30 AM Jan Skurovec ***@***.***>
wrote:

> self.device.type returns 'mps'
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/invoke-ai/InvokeAI/pull/1066#issuecomment-1276275684>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3EVLIQI5ODUDNDWE3Y2LWC3DOVANCNFSM6AAAAAARC6U4ZI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Written on my cell phone. Anything that seems odd is the fault of
auto-correct.

skurovec commented:
After fix switching models works. Images are consistent between models.

Vargol commented:
Yep, macOS is so lying about the free memory

```
NOTE: Redirects are currently not supported in Windows or MacOs.
>> GFPGAN Initialized
>> CodeFormer Initialized
>> ESRGAN Initialized
checking pop
8359503462.4
2147483648
>> Loading stable-diffusion-1.4 from models/ldm/stable-diffusion-v1/model.ckpt
   >> LatentDiffusion: Running in eps-prediction mode
   >> DiffusionWrapper has 859.52 M params.
   >> Making attention of type 'vanilla' with 512 in_channels
   >> Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
   >> Making attention of type 'vanilla' with 512 in_channels
>> Using more accurate float32 precision
>> Model loaded in 67.97s
>> Setting Sampler to k_lms

* Initialization done! Awaiting your command (-h for help, 'q' to quit)
invoke> !models
laion400m                 not loaded  Latent Diffusion LAION400M model
stable-diffusion-1.4          active  Stable Diffusion inference model version 1.4
waifu-1.3                 not loaded  Waifu 1.3
waifu-1.x                 not loaded  Waifu 1.3
invoke> !switch waifu-1.x
>> Caching model stable-diffusion-1.4 in system RAM
checking pop
3794085478.4
2147483648
>> Loading waifu-1.x from models/ldm/stable-diffusion-v1/model-waifu.ckpt
   >> LatentDiffusion: Running in eps-prediction mode
   >> DiffusionWrapper has 859.52 M params.
   >> Making attention of type 'vanilla' with 512 in_channels
   >> Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
   >> Making attention of type 'vanilla' with 512 in_channels
>> Using more accurate float32 precision
>> Model loaded in 84.83s
>> Setting Sampler to k_lms
invoke> !switch waifu-1.3
>> Caching model waifu-1.x in system RAM
checking pop
3696191078.4
2147483648
invoke> !models
laion400m                 not loaded  Latent Diffusion LAION400M model
stable-diffusion-1.4          cached  Stable Diffusion inference model version 1.4
waifu-1.3                     active  Waifu 1.3
waifu-1.x                     cached  Waifu 1.3
invoke> 

```

Vargol commented:
hold on

AVG_MODEL_SIZE=2.1*GIGS
DEFAULT_MIN_AVAIL=2*GIGS


        print(avail_memory + AVG_MODEL_SIZE)
        print(self.min_avail_mem)
        if avail_memory + AVG_MODEL_SIZE < self.min_avail_mem:

If you're using DEFAULT_MIN_AVAIL for self.min_avail_mem wouldn't  avail_memory have to be negative to pop
as AVG_MODEL_SIZE > DEFAULT_MIN_AVAIL
or am I having a brain fart....

lstein commented:
> hold on
> 
> AVG_MODEL_SIZE=2.1_GIGS DEFAULT_MIN_AVAIL=2_GIGS
> 
> ```
>     print(avail_memory + AVG_MODEL_SIZE)
>     print(self.min_avail_mem)
>     if avail_memory + AVG_MODEL_SIZE < self.min_avail_mem:
> ```
> 
> If you're using DEFAULT_MIN_AVAIL for self.min_avail_mem wouldn't avail_memory have to be negative to pop as AVG_MODEL_SIZE > DEFAULT_MIN_AVAIL or am I having a brain fart....

This is what comes from coding at 2 am, and also why it's so important to have multiple eyes on the code! It looks like I got this backward, and I'm surprised that it worked on my system. But I probably tested it backwards too.

Latest commit fixes this, and addresses other misc issues. Thanks for the help debugging on M1!

skurovec commented:
Latest changes working on my m1.

Vargol commented:
Same here , on my  8Gb M1
Models swapping, rendered images reflect the model, images consistent if you swap back to the model used for a previous render and render with same parameters, and its popping the models too now

lstein commented:
So would someone provide a code review approval so that I can move on to working on the next version of outpainting?

blessedcoolant commented:
Tested this out. Works as advertised. Just a question though: how many models are we caching? If the mem limit is hardcoded to 2GB, would this be an issue on lower end systems?

Any-Winter-4079 commented:
`generate.py` does not have `fix_func`. Does the merge take care of that? (I hope)

skurovec commented:
@Any-Winter-4079 PR https://github.com/invoke-ai/InvokeAI/pull/1056 was merged about same time this was created, so it is possible it from version before fix

lstein commented:
> `generate.py` does not have `fix_func`. Does the merge take care of that? (I hope)

The switch-models branch was made before `fix-func` came in, and when I merge to `development` this will be resolved.

Any-Winter-4079 commented:
Just started testing this. I'm getting
`RuntimeError: Placeholder storage has not been allocated on MPS device!` after trying to load a model that does not exist and then creating an image (with the current model).

lstein commented:
I can confirm this bug on a CUDA system. I'd tested out the scenario of loading a missing model file earlier in development and it was working, but more recent changes apparently broke it. I'll make another commit later today, and probably include code for interactively adding and editing stanzas in the `models.yaml` file.

Any-Winter-4079 commented:
Thanks for the fix. One thing is stable-diffusion-1.4 can't be renamed in `models.yml` or it fails.
`"stable-diffusion-1.4" is not a known model name. Please check your models.yaml file`

Besides that, it works well as far as I've tested it. By the way, is there correlation between sd-1.4 and waifu-1.3 in terms of seeds? I find it pretty fascinating.
<img width="785" alt="image" src="https://user-images.githubusercontent.com/50542132/195692109-69e0b3c9-384e-49ae-ba07-4b9cf2c4602d.png">

<img width="791" alt="image" src="https://user-images.githubusercontent.com/50542132/195692182-dd42b343-a54d-40f8-b8f9-ef1f88816740.png">

<img width="787" alt="image" src="https://user-images.githubusercontent.com/50542132/195692261-affc0065-0087-484e-bd76-461d3d119d7c.png">

Any-Winter-4079 commented:
I'm going to keep testing a bit more, with larger images, variations, etc.

lstein commented:
I've added a new commit that gives `ldm.invoke.model_cache` an API for importing and editing model config stanzas. I'm hoping this will be useful for the WebGUI. I've also added `!import_model` and `!edit_model` commands to the CLI. I'd appreciate if you could test this a bit and try to break it by giving bad values, etc.

Here are examples of how they work:

## !import_model    
<pre>
invoke> <b>!import_model models/ldm/stable-diffusion-v1/model-epoch08-float16.ckpt</b>
>> Model import in process. Please enter the values needed to configure this model:
    
Name for this model: <b>waifu-diffusion</b>
Description of this model: <b>Waifu Diffusion v1.3</b>
Configuration file for this model: <b>configs/stable-diffusion/v1-inference.yaml</b>
Default image width: <b>512</b>
Default image height: <b>512</b>
>> New configuration:
waifu-diffusion:
config: configs/stable-diffusion/v1-inference.yaml
description: Waifu Diffusion v1.3
height: 512
weights: models/ldm/stable-diffusion-v1/model-epoch08-float16.ckpt
width: 512
OK to import [n]? <b>y</b>
>> Caching model stable-diffusion-1.4 in system RAM
 >> Loading waifu-diffusion from models/ldm/stable-diffusion-v1/model-epoch08-float16.ckpt
       | LatentDiffusion: Running in eps-prediction mode
       | DiffusionWrapper has 859.52 M params.
       | Making attention of type 'vanilla' with 512 in_channels
       | Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
       | Making attention of type 'vanilla' with 512 in_channels
       | Using faster float16 precision
...etc
</pre>
    
## !edit_model
    
<pre>
invoke> <b>!edit_model waifu-diffusion</b>
>> Editing model waifu-diffusion from configuration file ./configs/models.yaml
description: <b>Waifu diffusion v1.4beta</b>
weights: models/ldm/stable-diffusion-v1/<b>model-epoch10-float16.ckpt</b>
config: configs/stable-diffusion/v1-inference.yaml
width: 512
height: 512
    
>> New configuration:
waifu-diffusion:
    config: configs/stable-diffusion/v1-inference.yaml
    description: Waifu diffusion v1.4beta
    weights: models/ldm/stable-diffusion-v1/model-epoch10-float16.ckpt
    height: 512
    width: 512
    
OK to change [n]? <b>y</b>
>> Caching model stable-diffusion-1.4 in system RAM
>> Loading waifu-diffusion from models/ldm/stable-diffusion-v1/model-epoch10-float16.ckpt
...
</pre>

lstein commented:
> Thanks for the fix. One thing is stable-diffusion-1.4 can't be renamed in `models.yml` or it fails. `"stable-diffusion-1.4" is not a known model name. Please check your models.yaml file` I guess it uses a hard-coded value as default to load the model on initialization.

Yeah, `stable-diffusion-1.4` is the default model name in the event the `--model` is not specified on the command line. Perhaps we should extend the `models.yaml` syntax to have an entry marked <b>default: True</b>. What do you think?

Thanks for letting me know about the correlation between SD and Waifu. I hadn't noticed that. It's quite cool.

Any-Winter-4079 commented:
> Perhaps we should extend the models.yaml syntax to have an entry marked default: True. What do you think?

That's probably a good idea.

> I've added a new commit that gives ldm.invoke.model_cache an API for importing and editing model config stanzas. I'm hoping this will be useful for the WebGUI. I've also added !import_model and !edit_model commands to the CLI. I'd appreciate if you could test this a bit and try to break it by giving bad values, etc.

Testing now.

lstein commented:
Good feedback. I knew about the problem with autocomplete not picking up the new model until you restart the script, but the rest is new to me. I'll add more error reporting as well as a delete option.

I think I'll merge in now and add the error checks in a subsequent PR.
